sql/core/src/test/scala/org/apache/spark/sql/MergeAsOfSuite.scala

```diff
@@ -37,107 +31,167 @@ class MergeAsOfSuite extends QueryTest with SharedSQLContext{
 
     checkAnswer(
       df1.mergeAsOf(df2, df1("time"), df2("time"), df1("id"), df2("id")),
       Seq(
-        Row(2001, 1, 1.0, null),
-        Row(2002, 1, 1.2, null),
-        Row(2001, 2, 1.1, null)
+        Row(new Timestamp(2001), 1, 1.0, null),
+        Row(new Timestamp(2002), 1, 1.2, null),
+        Row(new Timestamp(2001), 2, 1.1, null)
+      ))
+  }
+
+  test("merge_asof tolerance") {
+    val quotes = Seq(
+      (new Timestamp(23), "GOOG", 720.50, 720.93),
+      (new Timestamp(23), "MSFT", 51.95, 51.96),
+      (new Timestamp(30), "MSFT", 51.97, 51.98),
+      (new Timestamp(41), "MSFT", 51.99, 52.00),
+      (new Timestamp(48), "GOOG", 720.50, 720.93),
+      (new Timestamp(49), "AAPL", 97.99, 98.01),
+      (new Timestamp(72), "GOOG", 720.50, 720.88),
+      (new Timestamp(75), "MSFT", 52.01, 52.03)
+    ).toDF("time", "ticker", "bid", "ask")
+
+    val trades = Seq(
+      (new Timestamp(23), "MSFT", 51.95, 75),
+      (new Timestamp(38), "MSFT", 51.95, 155),
+      (new Timestamp(48), "GOOG", 720.77, 100),
+      (new Timestamp(48), "GOOG", 720.92, 100),
+      (new Timestamp(48), "AAPL", 98.00, 100)
+    ).toDF("time", "ticker", "price", "quantity")
+
+    checkAnswer(
+      trades.mergeAsOf(quotes, trades("time"), quotes("time"), trades("ticker"), quotes("ticker"), 2),
+      Seq(
+        Row(new Timestamp(23), "MSFT", 51.95, 75, 51.95, 51.96),
+        Row(new Timestamp(38), "MSFT", 51.95, 155, null, null),
+        Row(new Timestamp(48), "GOOG", 720.77, 100, 720.5, 720.93),
+        Row(new Timestamp(48), "GOOG", 720.92, 100, 720.5, 720.93),
+        Row(new Timestamp(48), "AAPL", 98.0, 100, null, null)
+      ))
+  }
+
+  test("merge_asof tolerance allow exact matches") {
+    val quotes = Seq(
+      (new Timestamp(23), "GOOG", 720.50, 720.93),
+      (new Timestamp(23), "MSFT", 51.95, 51.96),
+      (new Timestamp(30), "MSFT", 51.97, 51.98),
+      (new Timestamp(41), "MSFT", 51.99, 52.00),
+      (new Timestamp(48), "GOOG", 720.50, 720.93),
+      (new Timestamp(49), "AAPL", 97.99, 98.01),
+      (new Timestamp(72), "GOOG", 720.50, 720.88),
+      (new Timestamp(75), "MSFT", 52.01, 52.03)
+    ).toDF("time", "ticker", "bid", "ask")
+
+    val trades = Seq(
+      (new Timestamp(23), "MSFT", 51.95, 75),
+      (new Timestamp(38), "MSFT", 51.95, 155),
+      (new Timestamp(48), "GOOG", 720.77, 100),
+      (new Timestamp(48), "GOOG", 720.92, 100),
+      (new Timestamp(48), "AAPL", 98.00, 100)
+    ).toDF("time", "ticker", "price", "quantity")
+
+    checkAnswer(
+      trades.mergeAsOf(quotes, trades("time"), quotes("time"), trades("ticker"), quotes("ticker"), 10, false),
+      Seq(
+        Row(new Timestamp(23), "MSFT", 51.95, 75, null, null),
+        Row(new Timestamp(38), "MSFT", 51.95, 155, 51.97, 51.98),
+        Row(new Timestamp(48), "GOOG", 720.77, 100, null, null),
+        Row(new Timestamp(48), "GOOG", 720.92, 100, null, null),
+        Row(new Timestamp(48), "AAPL", 98.0, 100, null, null)
       ))
```

Comment:
We should also write a naive merge asof join algorithm to validate our implementation. This way we can generate test cases and have better coverage.